### ISSUE TITLE
Why does it consume so much memory?

### ISSUE THREAD

**Issue by LinB203** [webdataset/webdataset#256]
![image](https://user-images.githubusercontent.com/62638829/224202130-902438a6-c9c7-4a95-b23c-1ce9807f989b.png)

**Comment by LinB203** [comment]
`class Decode:
    def __init__(self, args=None):
        self.target_len = args.target_len
        self.freq = args.freq
        self.aug_roll_mag = args.aug_audio_roll_mag
        self.num_frames = args.num_frames
        self.tokenizer = CLIPTokenizerFast.from_pretrained(args.tokenizer_name, cache_dir=args.cache_dir)
        self.tokenizer.mask_token_id = 0
        self.max_seq_length = args.max_seq_length
        self.mlm_probability = args.mlm_probability
        self.use_cap = args.use_cap
        self.pretrain_clip = args.pretrain_clip

    def __call__(self, sample):
        # try:
        #     with warnings.catch_warnings(record=True) as w:
        video = self.decode_video([sample[f"frm{i}.jpg"] for i in range(self.num_frames)])


        audio = self.decode_audio(sample['pth'], self.target_len, self.freq)
        captions = [sample[f'cap{i}.txt'] for i in range(8)] if self.use_cap else None
        input_ids, labels, attention_mask = self.decode_text(sample['txt'], captions)

        video_new, audio_new = deepcopy(video), deepcopy(audio)
        input_ids_new, labels_new, attention_mask_new = deepcopy(input_ids), deepcopy(labels), deepcopy(attention_mask)
        del video, audio, input_ids, labels, attention_mask, captions
        gc.collect()
        return video_new, audio_new, input_ids_new, labels_new, attention_mask_new
        # except Exception as e:
        #     logging.info(f"Error occurred: {e}")
        #     return None

    def decode_video(self, frames):
        results = []
        for frame in frames:
            with io.BytesIO(frame) as stream:
                img = Image.open(stream)
                img.load()
                img = img.convert('RGB')
            result = np.asarray(img)
            results.append(torch.from_numpy(result))
        results = torch.stack(results, dim=0)
        frames = aug_frame(results, 224)
        frames_new = deepcopy(frames)
        del frames, results
        gc.collect()
        return frames_new

    def decode_audio(self, x, target_len=1024, freq=128):
        with io.BytesIO(x) as stream:
            fbank = torch.load(stream)

        n_frames = fbank.shape[0]
        assert freq == fbank.shape[1]
        p = target_len - n_frames

        # cut and pad
        if p > 0:
            m = torch.nn.ZeroPad2d((0, 0, 0, p))
            fbank = m(fbank)
        elif p < 0:
            fbank = fbank[0: target_len, :]
        fbank = (fbank - (-4.2677393)) / (4.5689974 * 2)
        fbank = fbank.unsqueeze(0)
        fbank_new = deepcopy(fbank)
        del fbank
        gc.collect()
        return fbank_new

    def decode_text(self, title, captions=None):
        title = title.decode("utf-8")
        if captions is not None:
            cap_pre = np.random.randint(0, 3)
            cpa_mid = np.random.randint(3, 5)
            cap_last = np.random.randint(5, 8)
            text = title + ', ' + ', '.join([captions[cap_pre].decode("utf-8"),
                                             captions[cpa_mid].decode("utf-8"),
                                             captions[cap_last].decode("utf-8")])
            text = whitespace_clean(basic_clean(text)).lower()
            del captions
            gc.collect()
        else:
            text = title

        text = self.tokenizer(text, padding='max_length', truncation=True, max_length=self.max_seq_length,
                              return_attention_mask=True, return_tensors="pt")
        input_ids = text.input_ids.clone()
        labels = input_ids.clone()
        if not self.pretrain_clip:
            input_ids, labels = self.mask_text(input_ids, self.tokenizer.vocab_size, labels)
        attention_mask = text.attention_mask
        input_ids, labels, attention_mask = input_ids.squeeze(), labels.squeeze(), attention_mask.squeeze()

        input_ids_new, labels_new, attention_mask_new = deepcopy(input_ids), deepcopy(labels), deepcopy(attention_mask)
        del input_ids, labels, attention_mask, title
        gc.collect()

        return input_ids_new, labels_new, attention_mask_new



    def mask_text(self, input_ids, vocab_size, targets):
        probability_matrix = torch.full(targets.shape, self.mlm_probability)
        masked_indices = torch.bernoulli(probability_matrix).bool()

        masked_indices[input_ids == self.tokenizer.bos_token_id] = False
        masked_indices[input_ids == self.tokenizer.eos_token_id] = False      # cls and pad is eof in clip

        targets[~masked_indices] = -100  # We only compute loss on masked tokens
        # 80% of the time, we replace masked input tokens with tokenizer.mask_token ([MASK])
        indices_replaced = torch.bernoulli(torch.full(input_ids.shape, 0.8)).bool() & masked_indices
        input_ids[indices_replaced] = self.tokenizer.mask_token_id
        # 10% of the time, we replace masked input tokens with random word
        indices_random = torch.bernoulli(torch.full(input_ids.shape, 0.5)).bool() & masked_indices & ~indices_replaced
        random_words = torch.randint(vocab_size, input_ids.shape, dtype=torch.long)
        input_ids[indices_random] = random_words[indices_random]
        # The rest of the time (10% of the time) we keep the masked input tokens unchanged
        return input_ids, targets

def get_wds_dataset(args, is_train=True, epoch=0, floor=False):
    input_shards = args.train_data
    assert input_shards is not None
    resampled = getattr(args, 'dataset_resampled', False) and is_train

    num_samples, num_shards = get_dataset_size(input_shards)
    if not num_samples:
        num_samples = args.train_num_samples
        if not num_samples:
            raise RuntimeError(
                'Currently, number of dataset samples must be specified for training dataset. '
                'Please specify via `--train-num-samples` if no dataset length info present.')

    shared_epoch = SharedEpoch(epoch=epoch)  # create a shared epoch store to sync epoch to dataloader worker proc

    def build_pipeline(input_shards):

        pipeline = [wds.SimpleShardList(input_shards)]

        # at this point we have an iterator over all the shards
        pipeline.extend([
            detshuffle2(
                bufsize=_SHARD_SHUFFLE_SIZE,
                initial=_SHARD_SHUFFLE_INITIAL,
                seed=args.seed,
                epoch=shared_epoch,
            ),
            wds.split_by_node,
            wds.split_by_worker,
            # at this point, we have an iterator over the shards assigned to each worker at each node
            tarfile_to_samples_nothrow,  # wds.tarfile_to_samples(handler=log_and_continue),
            wds.shuffle(
                bufsize=_SAMPLE_SHUFFLE_SIZE,
                initial=_SAMPLE_SHUFFLE_INITIAL,
            ),
            wds.map(Decode(args))
        ])
        return pipeline


    pipeline = build_pipeline(input_shards)

    pipeline.append(
        wds.batched(args.batch_size, partial=not is_train)
        # batched1(args.batch_size, partial=not is_train)
    )

    dataset = wds.DataPipeline(*pipeline)

    if not resampled:
        assert num_shards >= args.workers * args.world_size, 'number of shards must be >= total workers'
    # roll over and repeat a few samples to get same number of full batches on each node
    round_fn = math.floor if floor else math.ceil
    global_batch_size = args.batch_size * args.world_size
    num_batches = round_fn(num_samples / global_batch_size)
    num_workers = max(1, args.workers)
    num_worker_batches = round_fn(num_batches / num_workers)  # per dataloader worker
    num_batches = num_worker_batches * num_workers
    num_samples = num_batches * global_batch_size
    dataset = dataset.with_epoch(num_worker_batches)  # each worker is iterating over this

    dataloader = wds.WebLoader(
        dataset,
        batch_size=None,
        shuffle=False,
        num_workers=args.workers,
        persistent_workers=True,
    )

    dataloader.num_batches = num_batches
    dataloader.num_samples = num_samples

    return DataInfo(dataloader=dataloader, shared_epoch=shared_epoch)
`

**Comment by LinB203** [comment]
the pipeline is from open_clip ([https://github.com/mlfoundations/open_clip/blob/main/src/training/data.py](https://github.com/mlfoundations/open_clip/blob/main/src/training/data.py)), I just use my custom decode function. 
I don't know why it is consuming so much memory, how can I fix this problem? I have 200 .tar files, each .tar file contains 5k sets of data, each set has 16 frames, 1 image (.pth encoding), 1 title, 8 captions. i use 8 A100 GPUs for training, each GPU has a batchsize of 80 and 10 num workers.

**Comment by LinB203** [comment]
btw. I set the batchsize to 40, it's also failed, and the training time is longer than that of the batchsize 80.

**Comment by tmbdev** [comment]
WebDataset doesn't store or buffer any data unless you ask it to.

Most likely, your memory usage comes from the shuffle buffer. If your SHUFFLE_SIZE is 5000, that means that WDS keeps 5000 samples in memory. Try choosing a smaller size and see whether that reduces memory usage.

**Comment by vishaal27** [comment]
Hey @LinB203 did this resolve your issue? I also ran into the same issue where my CLIP training jobs get killed because of OOM, did you try reducing the buffer sizes and did that do any trick?
This is the exact issue I am facing : https://github.com/webdataset/webdataset/issues/325#issuecomment-2120638533
